src/uct/ib/rdmacm/rdmacm_cm_ep.c

```diff
@@ -149,15 +149,123 @@ ucs_status_t uct_rdmacm_cm_ep_conn_notify(uct_ep_h ep)
+uct_rdamcm_cm_ep_create_reserved_qpn(uct_rdmacm_cm_ep_t *cep,
+                                     struct uct_rdmacm_cm_qp_context *ctx)
+{
+    ucs_status_t status = UCS_OK;
+    struct uct_rdmacm_cm_reserved_qpn_blk *blk;
+    uint32_t qpns_per_obj;
+
+    qpns_per_obj = uct_rdmacm_cm_get_qpns_per_obj(ctx->reserved_qpn_ctx);
+
+    pthread_mutex_lock(&ctx->reserved_qpn_ctx.mutex);
+
+    blk = ucs_list_tail(&ctx->reserved_qpn_ctx.blk_list,
+                        struct uct_rdmacm_cm_reserved_qpn_blk, entry);
+    if (ucs_list_is_empty(&ctx->reserved_qpn_ctx.blk_list) ||
+        (blk->next_avail_qpn_offset >= qpns_per_obj)) {
+        status = uct_rdmacm_cm_reserved_qpn_blk_alloc(uct_rdmacm_cm_ep_get_cm(cep),
+                                                      cep->id->verbs, ctx, &blk);
+        if (status != UCS_OK) {
+            goto out;
+        }
+
+        ucs_list_add_tail(&ctx->reserved_qpn_ctx.blk_list, &blk->entry);
+    }
+
+    cep->qpn = blk->first_qpn + blk->next_avail_qpn_offset;
+    ucs_debug("created reserved qp_num: 0x%x on rdmacm_id %p",
+              cep->qpn, cep->id);
+    UCS_BITMAP_SET(blk->bmp, blk->next_avail_qpn_offset);
+    blk->next_avail_qpn_offset++;
+
+out:
+    pthread_mutex_unlock(&ctx->reserved_qpn_ctx.mutex);
+    return status;
+}
+
+/**
+ * Deallocate a reserved QPN. The FW object is destroyed only when all QPNs
+ * in this object were used and freed.
+ */
+static void uct_rdamcm_cm_ep_destroy_reserved_qpn(uct_rdmacm_cm_ep_t *cep)
+{
+    struct uct_rdmacm_cm_qp_context *ctx;
+    struct uct_rdmacm_cm_reserved_qpn_blk *blk;
+    uint32_t qpns_per_obj;
+    bool found;
+    ucs_status_t status;
+
+    if (!(cep->flags & UCT_RDMACM_CM_EP_QPN_CREATED)) {
+        return;
+    }
+
+    ucs_assert_always(cep->qpn != 0);
+
+    ucs_debug("cm ep destroy reserved qpn on rdmacm_id %p", cep->id);
+
+    status = uct_rdmacm_cm_get_qp_context(uct_rdmacm_cm_ep_get_cm(cep),
+                                          cep->id->verbs, &ctx);
+    if (status != UCS_OK) {
+        ucs_error("cm ep destroy reserved qpn failed on rdmacm_id %p: failed get qp context",
+                  cep->id);
+        return;
+    }
+
+    qpns_per_obj = uct_rdmacm_cm_get_qpns_per_obj(ctx->reserved_qpn_ctx);
+
+    pthread_mutex_lock(&ctx->reserved_qpn_ctx.mutex);
+
+    found = false;
+    ucs_list_for_each(blk, &ctx->reserved_qpn_ctx.blk_list, entry) {
+        if ((cep->qpn >= blk->first_qpn) &&
+            (cep->qpn < blk->first_qpn + qpns_per_obj)) {
+            found = true;
+            break;
+        }
+    }
+
+    if (!found || !UCS_BITMAP_GET(blk->bmp, cep->qpn - blk->first_qpn)) {
+        status = UCS_ERR_INVALID_PARAM;
+        goto out;
+    }
+
+    UCS_BITMAP_UNSET(blk->bmp, cep->qpn - blk->first_qpn);
+    if ((blk->next_avail_qpn_offset >= qpns_per_obj) &&
+        (UCS_BITMAP_IS_ZERO_INPLACE(&blk->bmp))) {
+        ucs_list_del(&blk->entry);
+        uct_rdmacm_cm_reserved_qpn_blk_free(blk);
```

Comment:
move ucs_list_del into uct_rdmacm_cm_reserved_qpn_blk_free() and rename it to uct_rdmacm_cm_reserved_qpn_blk_destroy()

Reply:
@yosefe Fixed.